src/proto/streams/streams.rs

```diff
@@ -951,6 +964,16 @@ fn drop_stream_ref(inner: &Mutex<Inner>, key: store::Key) {
 
     let actions = &mut me.actions;
 
+    // If the stream is not referenced and it is already
+    // closed (does not have to go through logic below
+    // of canceling the stream), we should notify the task
+    // (connection) so that it can close properly
+    if stream.ref_count == 0 && stream.is_closed() {
+        if let Some(ref task) = actions.task.as_ref() {
```

Comment:
This should probably be taken (`actions.task.take()`) as well. At least, that's the pattern that has been done through the lib whenever notifying on an `Option<Task>`.